File: qqzeng-phone-3.0/c++/PhoneSearchBest.cpp
#include "PhoneSearchBest.h"
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <unordered_map>
#include <locale>
#include <codecvt>
#include <algorithm>

using namespace std;





PhoneSearchBest* PhoneSearchBest::instance = nullptr;

// 单例模式获取 PhoneSearchBest 实例
PhoneSearchBest* PhoneSearchBest::GetInstance()
{
	if (!instance)
		instance = new PhoneSearchBest();
	return instance;
}
PhoneSearchBest::PhoneSearchBest() {
	LoadDat();
}
PhoneSearchBest::~PhoneSearchBest() {}

void PhoneSearchBest::LoadDat() {
	string dataPath = "qqzeng-phone-3.0.dat";//项目根目录

	std::locale::global(std::locale("zh_CN.UTF-8"));

	std::vector<unsigned char> bytes;
	std::ifstream file(dataPath, std::ios::binary);
	if (!file) {
		std::cerr << "Can't open the file." << std::endl;
		return;
	}

	file.seekg(0, std::ios::end);
	std::streampos fileSize = file.tellg();
	bytes.resize(fileSize);

	file.seekg(0, std::ios::beg);
	file.read((char*)&bytes[0], fileSize);



	int pref_size, desc_length, isp_length, phone_size;
	pref_size = BytesToInt<uint32_t>(bytes, 0);
	phone_size = BytesToInt<uint32_t>(bytes, 4);
	desc_length = BytesToInt<uint32_t>(bytes, 8);
	isp_length = BytesToInt<uint32_t>(bytes, 12);
	


	int head_length = 20;
	int start_index = head_length + desc_length + isp_length;


	string descString = BytesToUTF8String(bytes, head_length, desc_length);
	string ispString = BytesToUTF8String(bytes, head_length + desc_length, isp_length);


	addrArr = split(descString, '&');


	ispArr = split(ispString, '&');







	for (int m = 0; m < pref_size; m++)
	{
		int i = m * 7 + start_index;
		int pref = BytesToInt<uint8_t>(bytes, i);
		int index = BytesToInt<uint32_t>(bytes, i + 1);
		int length = BytesToInt<uint16_t>(bytes, i + 5);

		for (int n = 0; n < length; n++)
		{
			int p = (int)(start_index + pref_size * 7 + (n + index) * 4);
			int suff = BytesToInt<uint16_t>(bytes, p);
			int addrispIndex = BytesToInt<uint16_t>(bytes, p + 2);
			phone2D[pref][suff] = addrispIndex;
		}


	}

	file.close();
}

std::string PhoneSearchBest::BytesToUTF8String(const std::vector<unsigned char>& bytes, int pos, int len)
{
	std::string str(bytes.begin() + pos, bytes.begin() + pos + len);
	return str;
}

std::vector<std::string> PhoneSearchBest::split(std::string str, char delimiter) {

	std::vector<std::string> result;
	std::stringstream ss(str);
	std::string item;
	while (std::getline(ss, item, delimiter)) {
		result.push_back(item);
	}
	return result;
}


template <typename T>
T PhoneSearchBest::BytesToInt(const std::vector<unsigned char>& bytes, std::size_t m) {
	T result = 0;
	for (std::size_t i = 0; i < sizeof(T); ++i) {
		result |= (T(bytes[m + i]) << (i * 8));
	}
	return result;
}

// 查询方法
string PhoneSearchBest::Query(std::string  phoneNum) {
	int prefix = std::stoi(phoneNum.substr(0, 3));
	int suffix = std::stoi(phoneNum.substr(3, 4));
	int addrisp_index = phone2D[prefix][suffix];
	if (addrisp_index == 0) {
		return "";
	}
	return addrArr[addrisp_index / 100] + "|" + ispArr[addrisp_index % 100];
}

File: qqzeng-phone-3.0/dotnet/PhoneSearchBest.cs
    internal class PhoneSearchBest
    {
        private static readonly Lazy<PhoneSearchBest> lazy = new Lazy<PhoneSearchBest>(() => new PhoneSearchBest());
        public static PhoneSearchBest Instance { get { return lazy.Value; } }
        private PhoneSearchBest()
        {
            LoadDat();

        }

        private byte[] data;
        private long[,] phone2D;
        private string[] addrArr;
        private string[] ispArr;

        /// <summary>
        /// 初始化二进制dat数据
        /// </summary>
        /// <param name="dataPath"></param>
        /// 


        private void LoadDat()
        {

            var datPath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, @"qqzeng-phone-3.0.dat");

            data = File.ReadAllBytes(datPath);

            var PrefSize = BitConverter.ToUInt32(data, 0);

            var descLength = BitConverter.ToUInt32(data, 8);
            var ispLength = BitConverter.ToUInt32(data, 12);

            var PhoneSize = BitConverter.ToUInt32(data, 4);
            var verNum = BitConverter.ToUInt32(data, 16);

            var headLength = 20;
            int startIndex = (int)(headLength + descLength + ispLength);

            //内容数组        
            string descString = Encoding.UTF8.GetString(data, headLength, (int)descLength);
            addrArr = descString.Split('&');

            //运营商数组        
            string ispString = Encoding.UTF8.GetString(data, headLength + (int)descLength, (int)ispLength);
            ispArr = ispString.Split('&');


            phone2D = new long[200, 10000];
            for (var m = 0; m < PrefSize; m++)
            {
                int i = m * 7 + startIndex;
                int pref = data[i];
                int index = (int)BitConverter.ToUInt32(data, i + 1);
                int length = BitConverter.ToUInt16(data, i + 5);

                for (int n = 0; n < length; n++)
                {
                    int p = (int)(startIndex + PrefSize * 7 + (n + index) * 4);
                    var suff = BitConverter.ToUInt16(data, p);
                    var addrispIndex = BitConverter.ToUInt16(data, p + 2);
                    phone2D[pref, suff] = addrispIndex;
                }

            }



        }





        public string Query(string phone)
        {
            var prefix = Convert.ToInt32(phone.Substring(0, 3));//前缀
            var suffix = Convert.ToInt32(phone.Substring(3, 4));//后缀
            var addrispIndex = phone2D[prefix, suffix];
            if (addrispIndex == 0)
            {
                return "";
            }
            return addrArr[addrispIndex / 100] + "|" + ispArr[addrispIndex % 100];

        }




    }

    /*
    （调用例子）：    
   string result = PhoneSearchBest.Instance.Query("号段|号码");
   --> result="省份|城市|区号|邮编|行政区划代码|运营商"
    */
